src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                    //  7. Expression and expression statement have differing first tokens, indicating explicit discard (for example, VB call statement)
+                    //  8. Special cases: Methods beloning to System.Threading.Interlocked and System.Collections.Immutable.ImmutableInterlocked.
+                    var expressionStatement = (IExpressionStatementOperation)context.Operation;
+                    var value = expressionStatement.Operation;
+                    if (value.Type == null ||
+                        value.Type.SpecialType == SpecialType.System_Void ||
+                        value.Type.SpecialType == SpecialType.System_Boolean ||
+                        value.ConstantValue.HasValue ||
+                        value is IAssignmentOperation ||
+                        value is IIncrementOrDecrementOperation ||
+                        value is IInvalidOperation ||
+                        value.Syntax.GetFirstToken() != expressionStatement.Syntax.GetFirstToken() ||
+                        (value is IInvocationOperation invocation &&
+                         (invocation.TargetMethod.ContainingType.OriginalDefinition == _symbolStartAnalyzer._interlockedType ||
+                          invocation.TargetMethod.ContainingType.OriginalDefinition == _symbolStartAnalyzer._immutableInterlockedType)))
+                    {
+                        return;
+                    }
+
+                    // "Expression value is never used"
+                    var properties = s_propertiesMap[(_options.UnusedValueExpressionStatementPreference, isUnusedLocalAssignment: false, isRemovableAssignment: false)];
+                    var diagnostic = DiagnosticHelper.Create(s_expressionValueIsUnusedRule,
+                                                             value.Syntax.GetLocation(),
+                                                             _options.UnusedValueExpressionStatementSeverity,
+                                                             additionalLocations: null,
+                                                             properties);
+                    context.ReportDiagnostic(diagnostic);
+                }
+
+                private void AnalyzeDelegateCreationOrAnonymousFunction(OperationAnalysisContext operationAnalysisContext)
+                    => _hasDelegateCreationOrAnonymousFunction = true;
+
+                private void AnalyzeConversion(OperationAnalysisContext operationAnalysisContext)
+                {
+                    if (_hasConversionFromDelegateTypeToNonDelegateType)
+                    {
+                        return;
+                    }
+
+                    var conversion = (IConversionOperation)operationAnalysisContext.Operation;
+                    if (conversion.Operand.Type.IsDelegateType() &&
+                        !conversion.Type.IsDelegateType())
+                    {
+                        _hasConversionFromDelegateTypeToNonDelegateType = true;
```

Comment:
this sectoin is missing comments explaining why this is an important case to care about.  when reading this i say: ok... a delegate type was converted to a non-delegate (i guess 'object'?).  Why would that matter?  It feels like many cases were dsicovered, but there isn't any effort put into explaining what they are or why they matter.

Reply:
Added comments to the field declaration, with better refactoring on the code flow/explanations.

---
In reply to: [230605343](https://github.com/dotnet/roslyn/pull/30777#discussion_r230605343) [](ancestors = 230605343)